lib/Listener/AMembershipListener.php

```diff
@@ -0,0 +1,119 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Talk\Listener;
+
+use OCA\Circles\Api\v1\Circles;
+use OCA\Talk\Exceptions\ParticipantNotFoundException;
+use OCA\Talk\Manager;
+use OCA\Talk\Model\Attendee;
+use OCA\Talk\Participant;
+use OCA\Talk\Room;
+use OCA\Talk\Service\ParticipantService;
+use OCP\App\IAppManager;
+use OCP\EventDispatcher\IEventListener;
+use OCP\IGroupManager;
+use OCP\IUser;
+
+abstract class AMembershipListener implements IEventListener {
+
+	/** @var Manager */
+	protected $manager;
+	/** @var IAppManager */
+	protected $appManager;
+	/** @var IGroupManager */
+	protected $groupManager;
+	/** @var ParticipantService */
+	protected $participantService;
+
+	public function __construct(Manager $manager,
+								IAppManager $appManager,
+								IGroupManager $groupManager,
+								ParticipantService $participantService) {
+		$this->manager = $manager;
+		$this->appManager = $appManager;
+		$this->groupManager = $groupManager;
+		$this->participantService = $participantService;
+	}
+
+	protected function removeFromRoomsUnlessStillLinked(array $rooms, IUser $user): void {
+		$rooms = $this->filterRoomsWithOtherGroupMemberships($rooms, $user);
+		$rooms = $this->filterRoomsWithOtherCircleMemberships($rooms, $user);
+
+		foreach ($rooms as $room) {
+			try {
+				$participant = $room->getParticipant($user->getUID());
+				$participantType = $participant->getAttendee()->getParticipantType();
+				if ($participantType === Participant::USER) {
+					$this->participantService->removeUser($room, $user, Room::PARTICIPANT_REMOVED);
+				}
+			} catch (ParticipantNotFoundException $e) {
+			}
+		}
+	}
+
+	protected function filterRoomsWithOtherGroupMemberships(array $rooms, IUser $user): array {
+		$userGroupIds = $this->groupManager->getUserGroupIds($user);
+
+		$furtherMemberships = [];
+		foreach ($userGroupIds as $groupId) {
+			$groupRooms = $this->manager->getRoomsForActor(Attendee::ACTOR_GROUPS, $groupId);
+			foreach ($groupRooms as $room) {
+				$furtherMemberships[$room->getId()] = true;
+			}
+		}
+
+		return array_filter($rooms, static function (Room $room) use ($furtherMemberships) {
+			// Only delete from rooms where the user is not member via another group
+			return !isset($furtherMemberships[$room->getId()]);
+		});
+	}
+
+	protected function filterRoomsWithOtherCircleMemberships(array $rooms, IUser $user): array {
+		if (!$this->appManager->isEnabledForUser('circles', $user)) {
+			\OC::$server->getLogger()->error('Circles not enabled', ['app' => 'nickv']);
+			return $rooms;
+		}
+
+		try {
+			/** @var Circles $circlesApi */
+			$circles = Circles::joinedCircles($user->getUID());
```

Comment:
@ArtificialOwl

1. Is there a better way to get all circles of a user as this method is deprecated?
2. I guess I also need the parent circles of each to see if any of those is a member in a room. What's the API call for that?

Reply:
Done